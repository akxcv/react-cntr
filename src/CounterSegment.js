import React from 'react'
import ReactDOMServer from 'react-dom/server'
import AnimatedCounterDigit from './AnimatedCounterDigit'
import StaticCounterDigit from './StaticCounterDigit'
const { arrayOf, objectOf, number, string, any, func } = React.PropTypes

/**
 * @type {Object}
 * Maximum decimal values for available periods' numbers.
 */
const PERIOD_LIMITS = {
  'seconds': 59,
  'minutes': 59,
  'hours': 23,
  'days': 0
}

/**
 * Counter segment component.
 * @example
 * <CounterSegment
 *   digits={['0', '0']}
 *   period='days'
 *   radix={10}
 *   direction='down'
 *   easingDuration={300}
 *   digitMap={{ '0' => 'o' }}
 *   digitWrapper={(digit) => digit}
 * />
 */
class CounterSegment extends React.Component {
  /**
   * @property {string[]} digits - digits to display
   * @property {string} period
   * @property {number} radix
   * @property {string} direction - counting direction
   * @property {string} easingFunction - easing function to use for digit transitions
   * @property {number} easingDuration - duration of digit transitions
   * @property {Object} digitMap - a map for transforming particular digits
   * @property {function(digit: number)} digitWrapper - a function for wrapping mapped digits
   */
  static propTypes = {
    digits: arrayOf(string).isRequired,
    period: string.isRequired,
    radix: number.isRequired,
    direction: string.isRequired,
    easingFunction: string,
    easingDuration: number.isRequired,
    digitMap: objectOf(any).isRequired,
    digitWrapper: func.isRequired
  }

  /**
   * constructor
   * @param {Object} props
   */
  constructor (props) {
    super(props)

    /**
     * Calculates height of a single digit.
     */
    const testDigit = document.createElement('div')
    testDigit.innerHTML = ReactDOMServer.renderToString(this.props.digitWrapper('0'))
    document.getElementsByTagName('body')[0].appendChild(testDigit)

    /**
     * @type {Object}
     * @property {number} digitHeight - height of a single digit
     */
    this.state = {
      digitHeight: testDigit.clientHeight
    }
    testDigit.remove()
  }

  /**
   * Gets maximum value for period's digit with account for radix.
   * Used for building digit lanes in {@link AnimatedCounterDigit}.
   * @param {number} index - digit's index in number
   * @return {number} maxValue
   */
  getMaxValue (index) {
    const maxValue = PERIOD_LIMITS[this.props.period]
    if (!maxValue) return this.props.radix - 1

    const maxValueString = maxValue.toString(this.props.radix)
    const maxDigitPos = maxValueString.length
    return (index === this.props.digits.length - maxDigitPos)
      ? parseInt(maxValueString[0])
      : this.props.radix - 1
  }

  /**
   * Maps digits to corresponding components according to easing function.
   * @return {ReactElement[]}
   */
  buildDigits () {
    return this.props.digits.map((digit, index) => {
      if (this.props.easingFunction) {
        return (
          <AnimatedCounterDigit
            key={index}
            digit={digit}
            maxValue={this.getMaxValue(index)}
            radix={this.props.radix}
            direction={this.props.direction}
            easingFunction={this.props.easingFunction}
            easingDuration={this.props.easingDuration}
            height={this.state.digitHeight}
            digitMap={this.props.digitMap}
            digitWrapper={this.props.digitWrapper}
          />
        )
      } else {
        return (
          <StaticCounterDigit
            key={index}
            digit={digit}
            radix={this.props.radix}
            height={this.state.digitHeight}
            digitMap={this.props.digitMap}
            digitWrapper={this.props.digitWrapper}
          />
        )
      }
    })
  }

  /**
   * Renders the segment.
   * @return {ReactElement} counter segment
   */
  render () {
    const style = {
      overflow: 'hidden',
      height: `${this.state.digitHeight}px`
    }
    return (
      <div className='rollex-segment'>
        <div className='rollex-digits' style={style}>
          {this.buildDigits()}
        </div>
        <div className='rollex-label'>
          {this.props.period}
        </div>
      </div>
    )
  }
}

export default CounterSegment
