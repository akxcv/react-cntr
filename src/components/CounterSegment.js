import React from 'react'
import PropTypes from 'prop-types'
import AnimatedCounterDigit from './AnimatedCounterDigit'
import StaticCounterDigit from './StaticCounterDigit'

/**
 * @type {Object}
 * Maximum decimal values for available periods' numbers.
 */
const PERIOD_LIMITS = {
  'seconds': 59,
  'minutes': 59,
  'hours': 23,
  'days': 0
}

/**
 * @property {string[]} digits - digits to display
 * @property {string} period
 * @property {number} radix
 * @property {string} direction - counting direction
 * @property {string} easingFunction - easing function to use for digit transitions
 * @property {number} easingDuration - duration of digit transitions
 * @property {Object} digitMap - a map for transforming particular digits
 * @property {function(digit: number)} digitWrapper - a function for wrapping mapped digits
 * @property {string} label
 */
CounterSegment.propTypes = {
  digits: PropTypes.arrayOf(PropTypes.string).isRequired,
  period: PropTypes.string.isRequired,
  radix: PropTypes.number.isRequired,
  direction: PropTypes.string.isRequired,
  easingFunction: PropTypes.string,
  easingDuration: PropTypes.number.isRequired,
  digitMap: PropTypes.object.isRequired,
  digitWrapper: PropTypes.func.isRequired,
  label: PropTypes.string
}

/**
 * Counter segment component.
 * @example
 * <CounterSegment
 *   digits={['0', '0']}
 *   period='days'
 *   radix={10}
 *   direction='down'
 *   easingDuration={300}
 *   digitMap={{ '0' => 'o' }}
 *   digitWrapper={(digit) => digit}
 * />
 */
export default function CounterSegment (props) {
  /**
   * Gets maximum value for period's digit with account for radix.
   * Used for building digit lanes in {@link AnimatedCounterDigit}.
   * @param {number} index - digit's index in number
   * @return {number} maxValue
   */
  function getMaxValue (index) {
    const maxValue = PERIOD_LIMITS[props.period]
    if (!maxValue) return props.radix - 1

    const maxValueString = maxValue.toString(props.radix)
    const maxDigitPos = maxValueString.length
    return (index === props.digits.length - maxDigitPos)
      ? parseInt(maxValueString[0])
      : props.radix - 1
  }

  /**
   * Maps digits to corresponding components according to easing function.
   * @return {ReactElement[]}
   */
  function buildDigits () {
    return props.digits.map((digit, index) => {
      if (props.easingFunction) {
        return (
          <AnimatedCounterDigit
            key={index}
            digit={digit}
            maxValue={getMaxValue(index)}
            radix={props.radix}
            direction={props.direction}
            easingFunction={props.easingFunction}
            easingDuration={props.easingDuration}
            digitMap={props.digitMap}
            digitWrapper={props.digitWrapper}
          />
        )
      } else {
        return (
          <StaticCounterDigit
            key={index}
            digit={digit}
            radix={props.radix}
            digitMap={props.digitMap}
            digitWrapper={props.digitWrapper}
          />
        )
      }
    })
  }

  return (
    <div className='rollex-segment'>
      <div className='rollex-digits' style={{ overflow: 'hidden' }}>
        {buildDigits()}
      </div>
      <div className='rollex-label'>
        {props.label}
      </div>
    </div>
  )
}
