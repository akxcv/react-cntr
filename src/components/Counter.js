import React from 'react'
import CounterSegment from './CounterSegment'
import CounterSegmentSeparator from './CounterSegmentSeparator'
import GlobalIntervals from '../helpers/globalIntervals'
import CounterBuilder from '../helpers/counterBuilder'
import NumberCalculator from '../helpers/numberCalculator'
const { number, string, bool, objectOf, object, oneOfType, oneOf, func, any } = React.PropTypes

/**
 * Main counter component.
 * @example
 * <Counter seconds={98} />
 */
export default class Counter extends React.Component {
  /**
   * @property {number} from - timestamp to count from
   * @property {number} to - timestamp to count to
   * @property {number} seconds - a number of seconds to count
   * @property {number} interval - update interval
   * @property {string} direction - direction to count in
   * @property {number|object} digits - number of digits per segment
   * @property {string} minPeriod - smallest period to create a segment for
   * @property {string} maxPeriod - largest period to create a segment for
   * @property {boolean} frozen - determines if counter is frozen
   * @property {boolean} syncTime - determines if counter should try to synchronize with client
   * time.
   * @property {string} easingFunction - easing function to use for rolling digits
   * @property {number} easingDuration - easing duration in milliseconds
   * @property {number} radix - numeral base to use
   * @property {Object} digitMap - a map to use when transforming digit symbols
   * @property {function(digit: number)} digitWrapper - a wrapping function for mapped digits
   * @property {Map<string, string>|function(period: string, number: number)} labels - a map or a
   * function that returns labels for given period and number
   * @property separator - what to separate segments with
   */
  static propTypes = {
    from: number,
    to: number,
    seconds: number,
    interval: number,
    direction: oneOf(['up', 'down']),
    digits: oneOfType([number, object]),
    minPeriod: string,
    maxPeriod: string,
    frozen: bool,
    syncTime: bool,
    easingFunction: string,
    easingDuration: number,
    radix: number,
    digitMap: object,
    digitWrapper: func,
    labels: oneOfType([objectOf(string), func]),
    separator: any
  }

  static defaultProps = {
    interval: 1000,
    frozen: false,
    direction: 'down',
    minPeriod: 'seconds',
    maxPeriod: 'days',
    easingFunction: null,
    easingDuration: 300,
    radix: 10,
    digitMap: {},
    digitWrapper: (digit) => <span>{digit}</span>
  }

  /**
   * constructor
   * @param {Object} props
   */
  constructor (props) {
    CounterBuilder.validateProps(props)
    super(props)

    /**
     * @type {object}
     * @property {number} timeDiff - current amount of time to count from in milliseconds
     * @property {number} initialTimeDiff - original time diff
     * @property {Map<string, number>} digits - a map from periods to theis segment sizes
     * @property {Map<string, number>} numbers - a map from periods to their corresponding numbers
     * @property {number} startTime - a timestamp of current moment
     * @property {number} from - timestamp to count from
     * @property {string[]} periods - an array of periods to create segments for
     */
    this.state = new CounterBuilder(props).buildInitialState()

    /**
     * Creates a function bound to "this"
     * for subscribing to and unsubscribing from global Rollex tick event.
     */
    this.boundTick = this.tick.bind(this)
  }

  componentDidMount () {
    if (!this.props.frozen) this.start()
  }

  componentWillUnmount () {
    if (!this.props.frozen && !this.stopped) this.stop()
  }

  /**
   * Handles prop updates.
   * @param {Object} nextProps - new props
   */
  componentWillReceiveProps (nextProps) {
    if (this.props.frozen !== nextProps.frozen) {
      if (nextProps.frozen) {
        this.stop()
      } else {
        this.start()
      }
    }
  }

  /**
   * Starts the countdown.
   */
  start () {
    this.unsubscribe = GlobalIntervals.subscribe(this.props.interval, this.boundTick)
    this.stopped = false
  }

  /**
   * Pauses the countdown.
   */
  stop () {
    if (!this.unsubscribe) return
    this.unsubscribe()
    this.stopped = true
  }

  /**
   * Handles counter ticks.
   */
  tick () {
    const newTimeDiff = this.getTimeDiff()
    if (newTimeDiff < 0) {
      return this.stop()
    }

    const timestamp = (this.props.direction === 'down')
      ? newTimeDiff
      : this.state.initialTimeDiff - newTimeDiff

    /**
     * @type {object}
     * @property {number} timeDiff - current amount of time to count from in milliseconds
     * @property {Object} numbers - a map from periods to their corresponding numbers
     */
    this.setState({
      timeDiff: newTimeDiff,
      numbers: NumberCalculator.calculateNumbers(this.state.periods, timestamp)
    })
  }

  /**
   * Gets current amount of time to count from.
   * @return {number} timestamp
   */
  getTimeDiff () {
    if (this.props.syncTime) {
      return this.props.to - this.state.from - new Date().getTime() + this.state.startTime
    } else {
      return this.state.timeDiff - this.props.interval
    }
  }

  /**
   * Gets correct digits for given period accounting for counter's radix and periods's segment size.
   * @param {string} period
   * @return {string[]} digits
   */
  getDigits (period) {
    var number = this.state.numbers[period]
    const digits = this.state.digits[period]
    const radix = this.props.radix

    if (digits && number >= Math.pow(radix, digits)) {
      var maxValueArray = []
      for (let i = 0; i < digits; i++) maxValueArray.push((radix - 1).toString())
      return maxValueArray
    }

    number = number.toString(radix)
    var zeroArray = []
    for (let i = 0; i < digits - number.length; i++) zeroArray.push('0')
    return (zeroArray.join('') + number).split('')
  }

  /**
   * Gets a label for given period and number.
   * @param {string} period
   * @param {number} number
   * @return {string} label
   */
  getLabel (period, number) {
    if (!this.props.labels) {
      return period
    } else if (typeof this.props.labels === 'function') {
      return this.props.labels(period, number)
    } else {
      return this.props.labels[period]
    }
  }

  /**
   * Gets CSS classes for the counter.
   * @return {string} cssClasses
   */
  getCSSClasses () {
    const type = this.props.easingFunction ? 'animated' : 'static'
    var cssClasses = `rollex rollex-${type}`
    if (this.props.frozen) cssClasses += ' rollex-frozen'
    return cssClasses
  }

  /**
   * Renders the counter.
   * @return {ReactElement} counter
   */
  render () {
    const segments = this.state.periods.map((period, index) => {
      const separator = index < this.state.periods.length - 1
        ? <CounterSegmentSeparator content={this.props.separator} />
        : null
      return (<div key={index} className='rollex-segment-wrapper'>
        <CounterSegment
          period={period}
          digits={this.getDigits(period)}
          radix={this.props.radix}
          direction={this.props.direction}
          easingFunction={this.props.easingFunction}
          easingDuration={this.props.easingDuration}
          digitMap={this.props.digitMap}
          digitWrapper={this.props.digitWrapper}
          label={this.getLabel(period, this.state.numbers[period])}
        />
        {separator}
      </div>)
    })
    return (
      <div className={this.getCSSClasses()}>
        {segments}
      </div>
    )
  }
}
