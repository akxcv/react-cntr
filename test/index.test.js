import React from 'react'
import { shallow, mount } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import Counter from '../src/'
import CounterSegment from '../src/CounterSegment'

describe('initialization', function () {
  it('throws an error when options are provided incorrectly', function () {
    const _error = console.error
    console.error = jest.fn(() => null) // disable React PropTypes warnings

    expect(
      () => shallow(<Counter />)
    ).toThrowError('provide either "seconds" or "to"')
    expect(
      () => shallow(<Counter from={0} to={1} seconds={2} />)
    ).toThrowError('cannot use "to" and "from" with "seconds"')
    expect(
      () => shallow(<Counter from={0} />)
    ).toThrowError('provide either "seconds" or "to"')
    expect(
      () => shallow(<Counter from={0} seconds={2} />)
    ).toThrowError('cannot use "to" and "from" with "seconds"')
    expect(
      () => shallow(<Counter to={0} seconds={2} />)
    ).toThrowError('cannot use "to" and "from" with "seconds"')
    expect(
      () => shallow(<Counter from={1} to={0} />)
    ).toThrowError('"to" must be bigger than "from"')
    expect(
      () => shallow(<Counter seconds={-1} />)
    ).toThrowError('"seconds" must be greater than or equal to zero')
    expect(
      () => shallow(<Counter seconds={0} minDigits={0} />)
    ).toThrowError('"minDigits" must be positive')
    expect(
      () => shallow(<Counter seconds={0} minPeriod='sobaka' />)
    ).toThrowError('"minPeriod" must be one of: day, hour, minute, second')
    expect(
      () => shallow(<Counter seconds={0} maxPeriod='sobaka' />)
    ).toThrowError('"maxPeriod" must be one of: day, hour, minute, second')
    expect(
      () => shallow(<Counter seconds={0} radix={37} />)
    ).toThrowError('"radix" must be between 2 and 36')
    expect(
      () => shallow(<Counter seconds={0} digitWrapper={48} />)
    ).toThrowError('"digitWrapper" must be a function')
    expect(
      () => shallow(<Counter seconds={0} digitMap='111' />)
    ).toThrowError('"digitMap" must be an object')
    expect(
      () => shallow(<Counter to={0} direction='sobaka' />)
    ).toThrowError('"direction" must be either up or down')

    console.error = _error
  })

  it('initializes when options are correct', function () {
    expect(
      () => shallow(<Counter to={0} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter from={0} to={1} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={2} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={2} minDigits={1} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={2} maxDigits={3} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={2} minPeriod='minute' />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={2} maxPeriod='hour' />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter from={0} to={1} syncTime />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={0} syncTime />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={0} radix={12} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={0} digitWrapper={() => null} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter seconds={0} digitMap={{}} />)
    ).not.toThrow()
    expect(
      () => shallow(<Counter to={9} direction='up' />)
    ).not.toThrow()
  })
})

describe('rendering', function () {
  const to =
    (1000 * 60 * 60 * 24 * 200) +
    (1000 * 60 * 60 * 6) +
    (1000 * 60 * 35) +
    (1000 * 54)

  it('matches snapshot', function () {
    const component = shallow(<Counter from={0} to={1} />)
    const tree = shallowToJson(component)
    expect(tree).toMatchSnapshot()
  })

  it('renders a CounterSegment for each segment', function () {
    const component = shallow(<Counter from={0} to={1} />)
    expect(component.find(CounterSegment).length).toEqual(4)
  })

  it('passes digits correctly', function () {
    const component = shallow(<Counter from={0} to={to} />)
    const counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digits).toEqual(['2', '0', '0'])
    expect(counterSegments.at(1).props().digits).toEqual(['0', '6'])
    expect(counterSegments.at(2).props().digits).toEqual(['3', '5'])
    expect(counterSegments.at(3).props().digits).toEqual(['5', '4'])
  })

  it('works with "up" direction', function () {
    const component = shallow(<Counter from={0} to={to} direction='up' />)
    const counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digits).toEqual(['0', '0'])
    expect(counterSegments.at(1).props().digits).toEqual(['0', '0'])
    expect(counterSegments.at(2).props().digits).toEqual(['0', '0'])
    expect(counterSegments.at(3).props().digits).toEqual(['0', '0'])
  })

  test('minDigits', function () {
    var component = shallow(<Counter from={0} to={to} minDigits={1} />)
    var counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digits).toEqual(['2', '0', '0'])
    expect(counterSegments.at(1).props().digits).toEqual(['6'])
    expect(counterSegments.at(2).props().digits).toEqual(['3', '5'])
    expect(counterSegments.at(3).props().digits).toEqual(['5', '4'])

    component = shallow(<Counter from={0} to={to} minDigits={3} />)
    counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digits).toEqual(['2', '0', '0'])
    expect(counterSegments.at(1).props().digits).toEqual(['0', '0', '6'])
    expect(counterSegments.at(2).props().digits).toEqual(['0', '3', '5'])
    expect(counterSegments.at(3).props().digits).toEqual(['0', '5', '4'])
  })

  test('maxDigits', function () {
    var component = shallow(<Counter from={0} to={to} maxDigits={1} />)
    var counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digits).toEqual(['9'])
    expect(counterSegments.at(1).props().digits).toEqual(['6'])
    expect(counterSegments.at(2).props().digits).toEqual(['9'])
    expect(counterSegments.at(3).props().digits).toEqual(['9'])

    component = shallow(<Counter from={0} to={to} maxDigits={2} />)
    counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digits).toEqual(['9', '9'])
    expect(counterSegments.at(1).props().digits).toEqual(['0', '6'])
    expect(counterSegments.at(2).props().digits).toEqual(['3', '5'])
    expect(counterSegments.at(3).props().digits).toEqual(['5', '4'])
  })

  test('minPeriod', function () {
    const component = shallow(<Counter from={0} to={to} minPeriod='minute' />)
    const counterSegments = component.find(CounterSegment)
    expect(counterSegments.length).toBe(3)
    expect(counterSegments.at(2).props().period).toBe('minutes')
    expect(counterSegments.at(2).props().digits).toEqual(['3', '5'])
  })

  test('maxPeriod', function () {
    var component = shallow(<Counter from={0} to={to} maxPeriod='hour' />)
    var counterSegments = component.find(CounterSegment)
    expect(counterSegments.length).toBe(3)
    expect(counterSegments.at(0).props().period).toBe('hours')
    expect(counterSegments.at(0).props().digits).toEqual(['4', '8', '0', '6'])

    var component = shallow(<Counter from={0} to={to} maxPeriod='minute' />)
    var counterSegments = component.find(CounterSegment)
    expect(counterSegments.length).toBe(2)
    expect(counterSegments.at(0).props().period).toBe('minutes')
    expect(counterSegments.at(0).props().digits).toEqual(['2', '8', '8', '3', '9', '5'])

    var component = shallow(<Counter from={0} to={to} maxPeriod='second' />)
    var counterSegments = component.find(CounterSegment)
    expect(counterSegments.length).toBe(1)
    expect(counterSegments.at(0).props().period).toBe('seconds')
    expect(counterSegments.at(0).props().digits).toEqual(['1', '7', '3', '0', '3', '7', '5', '4'])
  })

  test('radix', function () {
    var component = shallow(<Counter from={0} to={to} radix={12} />)
    var counterSegments = component.find(CounterSegment)

    expect(counterSegments.at(0).props().digits).toEqual(['1', '4', '8'])
    expect(counterSegments.at(1).props().digits).toEqual(['0', '6'])
    expect(counterSegments.at(2).props().digits).toEqual(['2', 'b'])
    expect(counterSegments.at(3).props().digits).toEqual(['4', '6'])
  })

  it('passes digitWrapper to segments', function () {
    const digitWrapper = (digit) => digit + '0'
    const component = shallow(<Counter from={0} to={to} digitWrapper={digitWrapper} />)
    const counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digitWrapper).toEqual(digitWrapper)
    expect(counterSegments.at(1).props().digitWrapper).toEqual(digitWrapper)
    expect(counterSegments.at(2).props().digitWrapper).toEqual(digitWrapper)
    expect(counterSegments.at(3).props().digitWrapper).toEqual(digitWrapper)
  })

  it('passes digitMap to segments', function () {
    const digitMap = { '0': 'o' }
    const component = shallow(<Counter from={0} to={to} digitMap={digitMap} />)
    const counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().digitMap).toEqual(digitMap)
    expect(counterSegments.at(1).props().digitMap).toEqual(digitMap)
    expect(counterSegments.at(2).props().digitMap).toEqual(digitMap)
    expect(counterSegments.at(3).props().digitMap).toEqual(digitMap)
  })

  it('passes correct labels to segments', function () {
    const component = shallow(<Counter to={1} />)
    const counterSegments = component.find(CounterSegment)
    expect(counterSegments.at(0).props().label).toEqual('days')
    expect(counterSegments.at(1).props().label).toEqual('hours')
    expect(counterSegments.at(2).props().label).toEqual('minutes')
    expect(counterSegments.at(3).props().label).toEqual('seconds')

    const labelMap = {
      days: 'DD',
      hours: 'HH',
      minutes: 'MM',
      seconds: 'SS'
    }
    const labelMapComponent = shallow(<Counter to={1} labels={labelMap} />)
    const labelMapCounterSegments = labelMapComponent.find(CounterSegment)
    expect(labelMapCounterSegments.at(0).props().label).toEqual('DD')
    expect(labelMapCounterSegments.at(1).props().label).toEqual('HH')
    expect(labelMapCounterSegments.at(2).props().label).toEqual('MM')
    expect(labelMapCounterSegments.at(3).props().label).toEqual('SS')

    const labelFunc = function (period, number) {
      if (number % 10 === 1) {
        return period.slice(0, -1)
      } else {
        return period
      }
    }
    const labelFuncComponent = shallow(<Counter seconds={61} labels={labelFunc} />)
    const labelFuncCounterSegments = labelFuncComponent.find(CounterSegment)
    expect(labelFuncCounterSegments.at(0).props().label).toEqual('days')
    expect(labelFuncCounterSegments.at(1).props().label).toEqual('hours')
    expect(labelFuncCounterSegments.at(2).props().label).toEqual('minute')
    expect(labelFuncCounterSegments.at(3).props().label).toEqual('second')
  })
})

describe('state and props', function () {
  test('default state', function () {
    const from = (new Date()).getTime()
    const to = from +
      (1000 * 60 * 60 * 24 * 2) +
      (1000 * 60 * 60 * 6) +
      (1000 * 60 * 35) +
      (1000 * 54)
    const component = shallow(<Counter from={from} to={to} />)
    expect(component.state()).toMatchObject({
      timeDiff: to - from,
      numbers: {
        days: 2,
        hours: 6,
        minutes: 35,
        seconds: 54
      }
    })
  })

  test('default props', function () {
    const component = mount(<Counter from={0} to={1} />)
    expect(component.props()).toMatchObject({
      from: 0,
      to: 1,
      interval: 1000,
      minPeriod: 'second',
      maxPeriod: 'day',
      radix: 10,
      direction: 'down',
      frozen: false,
      easingFunction: null,
      easingDuration: 300
    })
  })

  it('allows to set props', function () {
    const digitWrapper = (digit) => <div>{digit}</div>
    const component = mount(
      <Counter from={10} to={20} interval={897}
        minDigits={3} maxDigits={4}
        minPeriod='minute' maxPeriod='hour'
        syncTime radix={8} direction='up'
        easingFunction='myEasingFn' easingDuration={123}
        digitMap={{ '0': 'o' }} digitWrapper={digitWrapper}
      />
    )
    expect(component.props()).toEqual({
      from: 10,
      to: 20,
      interval: 897,
      minDigits: 3,
      maxDigits: 4,
      minPeriod: 'minute',
      maxPeriod: 'hour',
      syncTime: true,
      easingFunction: 'myEasingFn',
      easingDuration: 123,
      radix: 8,
      frozen: false,
      direction: 'up',
      digitMap: { '0': 'o' },
      digitWrapper: digitWrapper
    })
  })
})

describe('counting', function () {
  beforeEach(function () {
    jest.clearAllTimers()
    jest.useFakeTimers()
    delete (window.rollexIntervals)
    // FIXME: the following line is required. It shouldn't be. Something's wrong with global state.
    mount(<Counter seconds={10000} />)
  })

  it('starts to count when mounted', function () {
    const component = mount(<Counter from={0} to={10000} />)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(5000)
    expect(component.state().numbers.seconds).toBe(5)
  })

  it('works with "seconds" prop', function () {
    const component = mount(<Counter seconds={10} />)
    expect(component.state().numbers.seconds).toBe(10)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(5000)
    expect(component.state().numbers.seconds).toBe(5)
  })

  it('works without "from" prop', function () {
    jest.spyOn(Date.prototype, 'getTime').mockImplementation(() => 78781234)
    const to = new Date().getTime() + 10000
    const component = mount(<Counter to={to} />)
    expect(component.state().timeDiff).toBe(10000)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(5000)
  })

  it('does not count when "frozen" is true', function () {
    const component = mount(<Counter seconds={10} frozen />)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(10000)
    expect(component.state().numbers.seconds).toBe(10)
  })

  it('stops to count when stopped', function () {
    const component = mount(<Counter from={0} to={10000} />)
    jest.runTimersToTime(5000)
    expect(component.state().numbers.seconds).toBe(5)

    component.instance().stop()

    jest.runTimersToTime(10000)
    expect(component.state().numbers.seconds).toBe(5)
  })

  it('stops to count when reached zero', function () {
    const component = mount(<Counter from={0} to={10000} />)
    jest.runTimersToTime(10000)
    expect(component.state().timeDiff).toBe(0)
    expect(component.state().numbers.seconds).toBe(0)

    jest.runTimersToTime(10000)
    expect(component.state().timeDiff).toBe(0)
    expect(component.state().numbers.seconds).toBe(0)
  })

  it('stops to count if frozen prop is updated to true', function () {
    const component = mount(<Counter from={0} to={10000} />)
    component.instance().stop = jest.genMockFunction()
    component.instance().forceUpdate()
    component.setProps({
      frozen: true
    }, () => expect(component.instance().stop.mock.calls.length).toBe(1))

    const frozenComponent = mount(<Counter from={0} to={10000} frozen />)
    frozenComponent.instance().stop = jest.genMockFunction()
    frozenComponent.instance().forceUpdate()
    frozenComponent.setProps({
      frozen: true
    }, () => expect(frozenComponent.instance().stop.mock.calls.length).toBe(0))
  })

  it('starts to count if frozen prop is updated to false', function () {
    const component = mount(<Counter from={0} to={10000} />)
    component.instance().start = jest.genMockFunction()
    component.instance().forceUpdate()
    component.setProps({
      frozen: false
    }, () => expect(component.instance().start.mock.calls.length).toBe(0))

    const frozenComponent = mount(<Counter from={0} to={10000} frozen />)
    frozenComponent.instance().start = jest.genMockFunction()
    frozenComponent.instance().forceUpdate()
    frozenComponent.setProps({
      frozen: false
    }, () => expect(frozenComponent.instance().start.mock.calls.length).toBe(1))
  })

  it('syncs time when syncTime is set', function () {
    const component = mount(<Counter from={0} to={10000} syncTime />)
    const newDate = component.state().startTime + 7000
    jest.spyOn(Date.prototype, 'getTime').mockImplementation(() => newDate)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(3000)
  })

  it('works with "up" direction', function () {
    const component = mount(<Counter seconds={10} direction='up' />)
    expect(component.state().timeDiff).toBe(10000)
    expect(component.state().numbers.seconds).toBe(0)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(5000)
    expect(component.state().numbers.seconds).toBe(5)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(0)
    expect(component.state().numbers.seconds).toBe(10)
    jest.runTimersToTime(5000)
    expect(component.state().timeDiff).toBe(0)
    expect(component.state().numbers.seconds).toBe(10)
  })
})

describe('CSS classes', function () {
  it('applies rollex and rollex-static classes to static counters', function () {
    const component = shallow(<Counter seconds={0} />)
    expect(component.find('.rollex').at(0).hasClass('rollex-static')).toBe(true)
  })

  it('applies rollex and rollex-animated classes to animated counters', function () {
    const component = shallow(<Counter seconds={0} easingFunction='ease-in' />)
    expect(component.find('.rollex').at(0).hasClass('rollex-animated')).toBe(true)
  })

  it('applies rollex-frozen class to frozen counters', function () {
    const staticComponent = shallow(<Counter seconds={0} frozen />)
    expect(staticComponent.find('.rollex').at(0).hasClass('rollex-frozen')).toBe(true)
    const animatedComponent = shallow(<Counter seconds={0} easingFunction='ease-in' frozen />)
    expect(animatedComponent.find('.rollex').at(0).hasClass('rollex-frozen')).toBe(true)
  })
})
