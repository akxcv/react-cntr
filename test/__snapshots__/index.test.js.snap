// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering matches snapshot 1`] = `
<div
  className="rollex rollex-static"
>
  <CounterSegment
    digitMap={Object {}}
    digitWrapper={[Function]}
    digits={
      Array [
        "0",
        "0",
      ]
    }
    easingDuration={300}
    easingFunction={null}
    period="days"
    radix={10}
  />
  <CounterSegment
    digitMap={Object {}}
    digitWrapper={[Function]}
    digits={
      Array [
        "0",
        "0",
      ]
    }
    easingDuration={300}
    easingFunction={null}
    period="hours"
    radix={10}
  />
  <CounterSegment
    digitMap={Object {}}
    digitWrapper={[Function]}
    digits={
      Array [
        "0",
        "0",
      ]
    }
    easingDuration={300}
    easingFunction={null}
    period="minutes"
    radix={10}
  />
  <CounterSegment
    digitMap={Object {}}
    digitWrapper={[Function]}
    digits={
      Array [
        "0",
        "0",
      ]
    }
    easingDuration={300}
    easingFunction={null}
    period="seconds"
    radix={10}
  />
</div>
`;
